decomp2d/CMakeLists.txt

```diff
@@ -0,0 +1,15 @@
+file(GLOB files_decomp decomp_2d.f90 glassman.f90)
+file(GLOB files_fft fft_generic.f90)
+set(SRCFILES ${files_decomp} ${files_fft})
+
+add_library(decomp2d STATIC ${SRCFILES})
+target_link_libraries(decomp2d PRIVATE MPI::MPI_Fortran)
+
+include(cmake/gfortran_flags.cmake)
+target_compile_options(decomp2d PRIVATE ${gfortran_flags})
+
+install(TARGETS decomp2d
+        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
```

Comment:
Do we need to care about installation location at the moment?

Reply:
Not really but what i did was for the full code that is why i had all this. I gave a quick look and seems ok to me. I have just to do a pull and check that everything compiles.

Reply:
we should also add this flag if gcc >= 10 `-fallow-argument-mismatch`. This is what i have at this line of the CMake https://github.com/xcompact3d/x3div/blob/eca75e84171e4aecb904be5b391bc58b7f91ae45/CMakeLists.txt#L85
Without this the code do not compile with new gcc

Reply:
The (long term) better solution would be to move to the mpi_f08 interfaces which have proper types, however tooling support is worse than rubbish so it's a non-starter currently

Reply:
I'm on Debian 11 with GCC 10.2
```
$ cmake ..
-- The Fortran compiler identification is GNU 10.2.1
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/f95 - skipped
-- Checking whether /usr/bin/f95 supports Fortran 90
-- Checking whether /usr/bin/f95 supports Fortran 90 - yes
-- Found MPI_Fortran: /usr/lib/x86_64-linux-gnu/libmpi_usempif08.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- MPI_Fortran_COMPILER found: /usr/bin/mpif90
-- Configuring done
-- Generating done
-- Build files have been written to: /home/thibault/scratch/repos/x3div/build
```
The branch compiles and runs. 

> Without this the code do not compile with new gcc

Are you talking about code introduced in other branches? I'm  not exactly clear on what `-fallow-argument-mismatch` is about.

Reply:
I think the error start from gcc 10.3 onwards. We had this issue on several clusters that run a relatively recent gcc. Below the esample from my MacOS with gcc 12.0, but it is the same on openSUSE with gcc11.4/5 
`rfj82982@ceg-rolfo-2 x3div % export FC=mpif90-mpich-mp
rfj82982@ceg-rolfo-2 build_minimal % cmake ../
-- The Fortran compiler identification is GNU 12.0.0
-- Checking whether Fortran compiler has -isysroot
-- Checking whether Fortran compiler has -isysroot - no
-- Checking whether Fortran compiler supports OSX deployment target flag
-- Checking whether Fortran compiler supports OSX deployment target flag - yes
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /opt/local/bin/mpif90-mpich-mp - skipped
-- Checking whether /opt/local/bin/mpif90-mpich-mp supports Fortran 90
-- Checking whether /opt/local/bin/mpif90-mpich-mp supports Fortran 90 - yes
-- Found MPI_Fortran: /opt/local/bin/mpif90-mpich-mp (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- MPI_Fortran_COMPILER found: /opt/local/bin/mpif90-mpich-mp
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/rfj82982/GIT_projects/x3div/build_minimal
rfj82982@ceg-rolfo-2 build_minimal % make 
Scanning dependencies of target decomp2d
[  6%] Building Fortran object decomp2d/CMakeFiles/decomp2d.dir/decomp_2d.f90.o
/Users/rfj82982/GIT_projects/x3div/decomp2d/halo_common.inc:167:15-15:

  167 | call MPI_IRECV(out(xs,ys,zs), 1, halo12, &
      |               1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/COMPLEX(4)).
/Users/rfj82982/GIT_projects/x3div/decomp2d/halo_common.inc:171:15:

  171 | call MPI_IRECV(out(xs,ye-level+1,zs), 1, halo12, &
      |               1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/COMPLEX(4)).
/Users/rfj82982/GIT_projects/x3div/decomp2d/halo_common.inc:175:16-16:
.....
`

This is to prevent the error on type mismatch on all MPI calls

Reply:
Okay, thanks for the clarification!